### PR TITLE
VIH-5827 Participant status component

### PR DESCRIPTION
**Pull request by rehmanab**
https://tools.hmcts.net/jira/browse/VIH-5827

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: VideoWeb/VideoWeb/ClientApp/src/app/services/api/video-web.service.ts [hunk 1]
```diff
@@ -125,6 +125,10 @@ export class VideoWebService implements IVideoWebApiService {
         return this.activeConferencesCache.get();
     }
 
+    getParticipantsByConferenceIdVho(conferenceId: string): Promise<ParticipantContactDetailsResponseVho[]> {
```
[the remark this answers is not in the document]
Comment (rehmanab) [reply VideoWeb/VideoWeb/ClientApp/src/app/services/api/video-web.service.ts:128]: It gets all the participants in the conference, they are EnRiched with contact details.. I will add Contact in the name

**Comment by rehmanab** [comment]
@shaed-parkar 
* Refactoring the cache should have gone into it's own PR
*Refactoring the eventhub client should have gone into its own PR

I think thats a pipe dream if the story is specific enough, in this case we had no idea at the beginning. Imagine how much work it would be to get half way through and then realise that the cache, eventhub needs changing, would be extra overhead splitting out into different PR's

File: VideoWeb/VideoWeb.Common/Caching/IConferenceCache.cs [hunk 2]
```diff
@@ -7,7 +7,8 @@ namespace VideoWeb.Common.Caching
 {
     public interface IConferenceCache
     {
-        Task AddConferenceToCache(ConferenceDetailsResponse conferenceResponse);
-        Conference GetConference(Guid id);
+        Task AddConferenceAsync(ConferenceDetailsResponse conferenceResponse);
+        Task <Conference>GetOrAddConferenceAsync(Guid id, Func<Task<ConferenceDetailsResponse>> addConferenceDetailsFactory);
+        Task<Conference> GetConferenceAsync(Guid id);
```
Comment (shaed-parkar) [inline VideoWeb/VideoWeb.Common/Caching/IConferenceCache.cs:12]: If we have an GetOrAddConferenceAsync, we can get rid of the GetConferenceAsync, both methods achieve the same thing

File: VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+describe('ParticipantContactDetails', () => {
+    it('should map participant info', () => {
+        const participants = new ConferenceTestData().getListOParticipantContactDetailsResponseVho(
+            'C7163972-A362-4167-8D33-77A64674B31C',
+            'MyVenue'
+        );
+        const participant = participants[0];
+        const p = new ParticipantContactDetails(participant);
+        p.judgeInAnotherHearing = true;
```
Comment (shaed-parkar) [inline VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts:12]: This is literally cheating. You've set line 12 to true and then expect it to be true. Better to set this value in the `ParticipantContactDetailsResponseVho` object

File: VideoWeb/VideoWeb/ClientApp/src/app/testing/mocks/MockChangesGuard.ts [hunk 4]
```diff
@@ -1,11 +1,11 @@
 import { CanDeactivate } from '@angular/router';
 import { CanDeactiveComponent } from 'src/app/shared/guards/changes.guard';
 export class MockChangesGuard implements CanDeactivate<CanDeactiveComponent> {
-    private _flag: boolean;
+    private flag: boolean;
```
Comment (shaed-parkar) [inline VideoWeb/VideoWeb/ClientApp/src/app/testing/mocks/MockChangesGuard.ts:4]: It doesn't look like we use the ChangesGuard anywhere in this app, probably worth removing it

File: VideoWeb/VideoWeb/Controllers/ParticipantsController.cs [hunk 5]
```diff
@@ -92,12 +104,12 @@ public async Task<IActionResult> GetTestCallResultForParticipantAsync(Guid confe
...
         private async Task<Guid> GetIdForParticipantByUsernameInConference(Guid conferenceId, string username)
         {
-            var conference = _conferenceCache.GetConference(conferenceId);
+            var conference = await _conferenceCache.GetConferenceAsync(conferenceId);
             if (conference == null)
             {
                 var conferenceDetail = await _videoApiClient.GetConferenceDetailsByIdAsync(conferenceId);
-                await _conferenceCache.AddConferenceToCache(conferenceDetail);
-                conference = _conferenceCache.GetConference(conferenceId);
+                await _conferenceCache.AddConferenceAsync(conferenceDetail);
```
Comment (shaed-parkar) [inline VideoWeb/VideoWeb/Controllers/ParticipantsController.cs:111]: would the new getOrAddToConference method be best suited here

**Comment by shaed-parkar** [comment]
> @shaed-parkar
> 
> * Refactoring the cache should have gone into it's own PR
>   *Refactoring the eventhub client should have gone into its own PR
> 
> I think thats a pipe dream if the story is specific enough, in this case we had no idea at the beginning. Imagine how much work it would be to get half way through and then realise that the cache, eventhub needs changing, would be extra overhead splitting out into different PR's

Definitely not a pipe dream. I mentioned this before you started the work. These are seperate changes, though may be a consequence of the story, could be implemented indepenedently.

> @shaed-parkar
> 
> * Refactoring the cache should have gone into it's own PR
>   *Refactoring the eventhub client should have gone into its own PR
> 
> I think thats a pipe dream if the story is specific enough, in this case we had no idea at the beginning. Imagine how much work it would be to get half way through and then realise that the cache, eventhub needs changing, would be extra overhead splitting out into different PR's

It's not a pipe dream, I suggested the seperate branch for the eventhub and even offered to help. The conference cache changes are not relevant to this story. As a team we agreed to keep the changes as small as possible and my original comment lists out those possible smaller changes.

File: VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts (hunk 3 above)
Comment (rehmanab) [reply VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts:12]: Its not cheating

**Comment by rehmanab** [comment]
@shaed-parkar 

> > @shaed-parkar
> > 
> > * Refactoring the cache should have gone into it's own PR
> >   *Refactoring the eventhub client should have gone into its own PR
> > 
> > I think thats a pipe dream if the story is specific enough, in this case we had no idea at the beginning. Imagine how much work it would be to get half way through and then realise that the cache, eventhub needs changing, would be extra overhead splitting out into different PR's
> 
> Definitely not a pipe dream. I mentioned this before you started the work. These are seperate changes, though may be a consequence of the story, could be implemented indepenedently.
> 
> > @shaed-parkar
> > 
> > * Refactoring the cache should have gone into it's own PR
> >   *Refactoring the eventhub client should have gone into its own PR
> > 
> > I think thats a pipe dream if the story is specific enough, in this case we had no idea at the beginning. Imagine how much work it would be to get half way through and then realise that the cache, eventhub needs changing, would be extra overhead splitting out into different PR's
> 
> It's not a pipe dream, I suggested the seperate branch for the eventhub and even offered to help. The conference cache changes are not relevant to this story. As a team we agreed to keep the changes as small as possible and my original comment lists out those possible smaller changes.

@shaed-parkar 
1) Still a pipe dream, the cache changes are essential to this PR, because this is a performance improvement task no? So If the participant list needs cached data and all you have is a get, I will have to write if statements to get and check and then get from source. I thought we are supposed to improve code as we go? I had to change the models for the cache to get my data, so why no improve the cache while I'm there ?
2) Yes you offered a separate branch, but so you really want me to wait until 2 PRs are in until I carry on with my work ?
3) if your talking about separate PR's, why did you want all the linting stuff done which is not part of this story in this PR, is that not eligible for a separate PR ?
4) If you want smaller PR's, then you need to spec that there are 3 tasks to the story and not let the developer get to a point where is too much overhead to push out different PR's

File: VideoWeb/VideoWeb/Controllers/ParticipantsController.cs (hunk 5 above)
Comment (rehmanab) [reply VideoWeb/VideoWeb/Controllers/ParticipantsController.cs:111]: This should be a separate PR

File: VideoWeb/VideoWeb/ClientApp/src/app/testing/mocks/MockChangesGuard.ts (hunk 4 above)
Comment (rehmanab) [reply VideoWeb/VideoWeb/ClientApp/src/app/testing/mocks/MockChangesGuard.ts:4]: This should be a separate PR

File: VideoWeb/VideoWeb.Common/Caching/IConferenceCache.cs (hunk 2 above)
Comment (rehmanab) [reply VideoWeb/VideoWeb.Common/Caching/IConferenceCache.cs:12]: This should be a separate PR

File: VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-status-reader.spec.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import { ParticipantStatusReader } from './participant-status-reader';
+import { ParticipantStatus } from '../../services/clients/api-client';
+
+describe('ParticipantStatusReader', () => {
```
[the remark this answers is not in the document]
Comment (rehmanab) [reply VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-status-reader.spec.ts:4]: Good shout

File: VideoWeb/VideoWeb.Common/Caching/IConferenceCache.cs (hunk 2 above)
Comment (rehmanab) [reply VideoWeb/VideoWeb.Common/Caching/IConferenceCache.cs:12]: But yes, we should use this where ever we are getting from cache and if null, setting it

File: VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts (hunk 3 above)
Comment (rehmanab) [reply VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts:12]: good catch though

**Comment by rehmanab** [comment]
> I mentioned this before you started the work. These are seperate changes, though may be a consequence of the story, could be implemented indepenedently.

No you didn't

File: VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts (hunk 3 above)
Comment (rehmanab) [reply VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-contact-details.spec.ts]: done

File: VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-status-reader.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (rehmanab) [reply VideoWeb/VideoWeb/ClientApp/src/app/shared/models/participant-status-reader.spec.ts:4]: done

**Pull request merged**